### ISSUE TITLE
DRAM-sharded matmuls ND bad PCC when running on TG 

### ISSUE THREAD

**Issue by mikevin920** [tenstorrent/tt-metal#10673]
When running DRAM-sharded matmuls on all 32 devices and then concating the results back on host will ND produces garbage outputs

**Comment by tt-rkim** [comment]
@aliuTT @ubcheema do you guys know what's going on with this one?

**Comment by ubcheema** [comment]
Can you check if the system has correct galaxy fw on it.

**Comment by ttmchiou** [comment]
@mikevin920 can you post a github run with this error?

**Comment by mikevin920** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/10021434594/job/27700225744.  Is this galaxy fw correct?

**Comment by tt-rkim** [comment]
It doesn't print out the firmware there. Looks like it's using an older version of the reset script

@ttmchiou we should update the TGG, TG, and T3K reset scripts to print out SMI JSON information.

Otherwise, for this current work, someone will have to log in and check SMI. It's tough because the machine is in use all the time and we need to be careful with resets.

**Comment by mikevin920** [comment]
![image](https://github.com/user-attachments/assets/91afc694-bbc5-44fa-987f-90a8b50b5547)
This is the local machine firmware im seeing consistent issues with

**Comment by ubcheema** [comment]
This is the last stable Galaxy Fw release:
`/mnt/motor/syseng/bin/tt-flash/wh/mobo/galaxy_fw_7.14.C.0_2024-07-04-00b2b9f7.tar.gz`
It runs at 800 MHz with additional voltage margin to account for di/dt.

Tyler Colaco is working on a new release that will run Galaxy at 900 MHz.

@tt-rkim I am on vacation will 08/03. Please sync with Tyler on where you pick the release from.

It's better to just reprogram the machine that @mikevin920 is working on to make sure it has correct fw package on it.

**Comment by davorchap** [comment]
do we know is this is a sharded DRAM Matmul issue or a model / concat issue?

**Comment by tt-rkim** [comment]
@mikevin920 did you guys want to try all the galaxy tests (including this one) on an upgraded galaxy in the corporate network (aus glx etc.)?

**Comment by cglagovichTT** [comment]
My local machine (aus-glx-07) has this tt-smi output:

<img width="781" alt="image" src="https://github.com/user-attachments/assets/6f5c5de7-5f49-4926-af79-ab6e886aed6f">
<img width="1269" alt="image" src="https://github.com/user-attachments/assets/2e56456c-0412-4904-b06a-8ab6fb84d244">


I'm able to repro ND PCC on branch based off of main @ 374510603bfe4c52afbf34581b62c37da50e1576. 
My test was a Llama MLP unit test on TG. Running matmul_1d for 1000 iterations gave consistently good PCC. Replacing matmul_1d with dram sharded matmul caused ~1/2 of the iterations to output bad PCC (close to 0). 

<img width="1567" alt="image" src="https://github.com/user-attachments/assets/35e2b541-31aa-4ee1-8818-2e517e222ec1">


I also reproed by running this test in a loop. The result was bad PCC every few iterations.
```bash
pytest -svv tests/ttnn/multichip_unit_tests/test_multidevice_TG.py::test_galaxy_matmul_2d_fracture_dram_sharded
```

**Comment by davorchap** [comment]
fyi @yugaoTT 

Is bad PCC reproducible on N300/T3000 using the same DRAM sharded MM shape/config? Wondering if this is specific to Galaxy or a matmul bug?

**Comment by cglagovichTT** [comment]
Reproed on N300 single chip, saw deterministic PCC for 1000 iteratinos

**Comment by yugaoTT** [comment]
Can I get a repro procedure? I'll run on my side to see what happens. @cglagovichTT is it only happening at iteration 1000?

**Comment by cglagovichTT** [comment]
^ I only tested for 1000 iterations on single-chip and all iterations passed

So far we have only seen ND on TG systems with 32 chips running the matmul.

Repro:
Get a galaxy from IRD. aus-glx-06 is available now and has the right FW

```bash
pytest  tests/ttnn/multichip_unit_tests/test_multidevice_TG.py::test_galaxy_matmul_2d_fracture_dram_sharded
```

You'll have to remove the pytest.skip on that test locally - we disabled it in CI because of ND

**Comment by ubcheema** [comment]
What about running in 8 chips, on a T3K machine? 
Does that also pass?

**Comment by cglagovichTT** [comment]
I just ran and confirmed that running on 8 chips on a T3K machine passes

**Comment by davorchap** [comment]
TG on 1, 2, 4, 8, 16 devices is passing ?

**Comment by davorchap** [comment]
For the 1st level triage, a data collection step helps expedite the debug across the teams.
Can  you fill this out:

<img width="742" alt="image" src="https://github.com/user-attachments/assets/e2f683ca-b4d0-4e5b-956b-ffae7fbcc16e">

**Comment by cglagovichTT** [comment]
<img width="982" alt="image" src="https://github.com/user-attachments/assets/da0038f2-dec6-4382-ba32-432f8d7427d3">

**Comment by yugaoTT** [comment]
could it be caused by dram sharding of the weight tensor? `test_galaxy_matmul_2d_fracture_dram_sharded` is calling the dram sharded matmul version, but can also try `tests/tt_eager/python_api_testing/unit_testing/misc/test_matmul_dram_sharded.py::test_matmul_2d_in1_dram_sharded` which is calling the normal matmul_2d, but with sharded in1 tensor

**Comment by cglagovichTT** [comment]
@yugaoTT this ND for this matmul shape has only shown up for dram sharded weight tensors. We have seen no problems using matmul_1d for this matmul

**Comment by yugaoTT** [comment]
So we have two versions of matmul that both support dram sharded weight: 
matmul dram sharded (MatmulMultiCoreReuseMultiCastDRAMShardedProgramConfig) 
and matmul_2d (MatmulMultiCoreReuseMultiCastProgramConfig). 

I think in your test it's testing the first one right? the second one might also have problems if dram sharding is causing the problem.

**Comment by cglagovichTT** [comment]
Yes we are using the first program config in our tests

**Comment by cglagovichTT** [comment]
On TG:

I reproed ND PCC with the following configs:

1. matmul_2d on 1x8 coregrid, activation interleaved, weight dram sharded, output interleaved
2. matmul_2d on 1x8 coregrid, activation interleaved, weight interleaved, output interleaved

I confirmed that the matmul used the expected kernel `bmm_large_block_zm_fused_bias_activation.cpp`, so this is a different issue than #10936. This is the tracy for the failing matmul 
[ops_perf_results_2024_08_14_17_17_25.csv](https://github.com/user-attachments/files/16616912/ops_perf_results_2024_08_14_17_17_25.csv)

Running on smaller clusters than 4x8 makes the ND PCC less frequent, probably because the op is executed fewer times in total.

**Comment by cglagovichTT** [comment]
Updated the branch to automatically sweep 4 types of matmuls:
1. dram sharded matmul - act sharded, weight sharded
2. matmul_1d - act sharded, weight interleaved
3. matmul_2d - act interleaved, weight sharded
4. matmul_2d - act interleaved, weight interleaved

I sweep these matmuls with two different input shapes which are used in Llama. 7/8 tests fail due to non-determinism.

<img width="1495" alt="image" src="https://github.com/user-attachments/assets/c8d8bb49-62d7-4f00-854a-1c934628ac38">


Repro:
This repro will run these 8 combinations for 2000 iterations each. I expect 7/8 tests to fail due to ND PCC.
branch `cglagovich/10673_repro`
` pytest -svv tests/ttnn/multichip_unit_tests/test_multidevice_TG.py::test_galaxy_matmul_2d_fracture_dram_sharded`

**Comment by cglagovichTT** [comment]
Ran these same exact matmul shapes and configs on t3k for 1000 iterations each. All passed 
<img width="1583" alt="image" src="https://github.com/user-attachments/assets/9cbe76d4-e279-4529-ac74-f06639a18a8f">

**Comment by cglagovichTT** [comment]
@TT-BrianLiu @yugaoTT I'm assigning to you for further triage/debug. LMK if you need more information from me

**Comment by cglagovichTT** [comment]
Retried on aus-glx-01 because we believed ND could be a machine-specific issue. 
<img width="1582" alt="image" src="https://github.com/user-attachments/assets/98daa2b8-7d17-4d7d-b183-d6e9fe708325">

**Comment by cglagovichTT** [comment]
Set clock to 500MHz by removing `set_power_state` from `tt_silicon_driver::deassert_resets_and_set_power_state`. I re-ran the 4 previously failing tests: 
<img width="1575" alt="image" src="https://github.com/user-attachments/assets/67e2a815-44c7-4752-9ba4-b5448eadf6cf">

**Comment by yugaoTT** [comment]
this sounds similar to an issue that I met a long time ago, when debugging the di/dt problem, sometimes the pcc is also garbage even if it does not hang.

**Comment by cglagovichTT** [comment]
Running some experiments to see if a specific device consistently produces ND output. Early result: device (6,1) produced 29.0 where previous runs produced 29.5.
<img width="1589" alt="image" src="https://github.com/user-attachments/assets/7fd6a4aa-19bf-4f17-8266-edb092dd3a56">

**Comment by cglagovichTT** [comment]
On another aus-glx-01 run @ 900 MHz, these are the stats I've collected across 1000 runs for the failing configs:

<details><summary>dram_sharded</summary>

</details>

**Comment by cglagovichTT** [comment]
Another run for 2000 iterations:

dram sharded FF1
```
----------
Chip (4, 1) failed 43 times
Columns that failed:
Index: (0, 527), Diff: 0.5 failed 24 times
Index: (0, 167), Diff: 0.5 failed 18 times
Index: (0, 27), Diff: 0.25 failed 1 times
----------

----------
Chip (5, 0) failed 13 times
Columns that failed:
Index: (0, 527), Diff: 0.5 failed 10 times
Index: (0, 167), Diff: 0.5 failed 3 times
----------

----------
Chip (0, 1) failed 12 times
Columns that failed:
Index: (0, 27), Diff: 0.25 failed 7 times
Index: (0, 459), Diff: 0.5 failed 4 times
Index: (0, 527), Diff: 0.5 failed 1 times
----------

----------
Chip (2, 0) failed 1 times
Columns that failed:
Index: (0, 595), Diff: 0.5 failed 1 times
----------

----------
Chip (4, 3) failed 4 times
Columns that failed:
Index: (0, 167), Diff: 0.5 failed 3 times
Index: (0, 559), Diff: 0.5 failed 1 times
----------

----------
Chip (6, 1) failed 5 times
Columns that failed:
Index: (0, 27), Diff: 0.25 failed 2 times
Index: (0, 459), Diff: 0.5 failed 2 times
Index: (0, 387), Diff: 0.5 failed 1 times
----------

----------
Chip (0, 0) failed 10 times
Columns that failed:
Index: (0, 243), Diff: -2.0 failed 4 times
Index: (0, 203), Diff: -inf failed 4 times
Index: (0, 51), Diff: -inf failed 1 times
Index: (0, 107), Diff: -inf failed 1 times
----------

----------
Chip (4, 2) failed 2 times
Columns that failed:
Index: (0, 27), Diff: 0.25 failed 2 times
----------

----------
Chip (7, 1) failed 1 times
Columns that failed:
Index: (0, 167), Diff: 0.5 failed 1 times
----------

----------
Chip (5, 1) failed 1 times
Columns that failed:
Index: (0, 459), Diff: 0.5 failed 1 times
----------
```

**Comment by cglagovichTT** [comment]
@yugaoTT and I are starting to believe that this ND PCC is caused by a bit flip during transmission of a tile from DRAM -> L1. 

- The matmul of `ones @ rand` produces an output where each column of the output is the sum of a column of the weights
- Each datum in a column of the output is expected to be equal
- The above data shows that the difference between `expected` and `out` is always a power of 2 or sum of powers of 2

-

**Comment by yugaoTT** [comment]
if some pins failed between dram and chip (I assume it will be always giving 0 for that pin), and if weight are positive numbers, then we would alway have computed value smaller then original?

**Comment by cglagovichTT** [comment]
This is the data on the number of failures per chip for each of the 8 configurations I ran. There's a pattern, where chips (3,0) and (7,0) are most likely to fail on all cases. For dram sharded matmuls, there are 7 chips which fail >40% of the time.
[chip_failures.txt](https://github.com/user-attachments/files/16710895/chip_failures.txt)

**Comment by cglagovichTT** [comment]
I re-ran on only chip 0,0 and got 2154 failures in 10k iterations. This failure does not only occur in 32-chip workloads...

**Comment by cglagovichTT** [comment]
Discussion continued in slack https://tenstorrent.slack.com/archives/C07384DMYJC/p1724331544384049

TLDR; bank 0 and 1 (GDDR0) are the only banks that we have seen fail. 
We separated compute from data movement and showed that the weights are still getting corrupted.

**Comment by johanna-rock-tt** [comment]
What's the current status of this issue? Has this been fixed in the meantime?

**Comment by cglagovichTT** [comment]
Alex Buck has been looking into this on the hardware side. Last I heard, he was able to repro data corruption in their own DRAM tests, and he has been able to repro my test as well. No workaround yet.

**Comment by cglagovichTT** [comment]
Latest info from Alex: 
>It looks like it might be a voltage margin related issue. Boosting one of the DRAM related voltages mitigates the issue. I'm trying to test on our larger testing systems to make sure we don't break anything else prior to releasing a FW update

**Comment by abuckTT** [comment]
At the moment, we have a potential fix for that system by increasing one of the voltages, but when I tried applying it to some other systems to make sure it doesn't break anything new, new failures cropped up. I'm rerunning that test on our "golden" sample, because my previous test also had some known bad modules on it. I'm hoping that that system doesn't have any issues.

**Comment by abuckTT** [comment]
Latest tests did not pass with increased voltage. Next steps, sweep higher range of PHY voltage up to maximum.

**Comment by ttmtrajkovic** [comment]
@abuckTT,
Could you please update on this debug, what is the current status and steps? 
Also, is this only on Galaxy related problem or you think n300 and n150 cards could be affected.

milos

**Comment by abuckTT** [comment]
Hi,
This only appears to be a 4U Galaxy related problem due to marginally failing parts. N300 and N150 should be much more stable.

Increasing PHY voltage causes new issues on other working modules, so it doesn't appear to be viable solution moving forward. Next I want to look into the screening process used on the modules that are affected on the colo. Some may have been earlier samples built prior to more extensive screening, which allowed more marginally failing parts to pass. Later modules were tested with more stress testing. Likely solution is to replace the marginal parts with better screened ones.

**Comment by uaydonat** [comment]
Are there any 6U Galaxy's available for early testing?

4U Galaxy's will be deprecated, if we are confident this issue does not happen on 6U we might be able to lower the priority.

**Comment by abuckTT** [comment]
I parsed the data for all the modules on the galaxy systems. Looking through all the systems we have on cloud, `aus-glx-mgmt-01` to `-06` were all built before we did a logged burnin test and used an earlier version of module functional test. `aus-glx-mgmt-7` to `-15` and `g*glx0*` systems all ran through the burnin test and more recent functional test. Failures on systems `aus-glx-mgmt-01 `to `-06` are likely caused by parts that weren't sufficiently screened.
@cglagovichTT , do you have data on which systems you experienced the failures on?

**Comment by johanna-rock-tt** [comment]
Recently I ran a benchmark that was also reading from dram (but not the dram sharded matmuls test from this issue) and didn't see any accuracy issues. I was running this benchmark on `aus-glx-08`. But in the screenshot from above I see this test was run on aus-glx-15 (see Hostname).

**Comment by cglagovichTT** [comment]
I've tested aus-glx-01 and aus-glx-07 and saw the failure on both

**Comment by abuckTT** [comment]
So it looks like it's affecting 4U systems that also went through our burn-in test. We don't have any 6U systems that are stable for general testing yet, but I can try running some workloads on the one we have. @cglagovichTT could you share how I can run the workload on a single chip? We have direct PCI access to each chip. Previously I was using `pytest -svv tests/ttnn/multichip_unit_tests/test_multidevice_TG.py::test_galaxy_matmul_2d_fracture_dram_sharded`, but this assumed a TG setup.  If there's a way I can run the workload on a single chip, I can check if 6U also has this issue or not.

**Comment by cglagovichTT** [comment]
Hi Alex, you can change this line to (1,1) in order to run on a single chip
https://github.com/tenstorrent/tt-metal/blob/cglagovich/10673_repro/tests/ttnn/multichip_unit_tests/test_multidevice_TG.py#L179

Putting the repro instructions here for future reference:
- branch `cglagovich/10673_repro`
- `pytest -svv tests/ttnn/multichip_unit_tests/test_multidevice_TG.py::test_galaxy_matmul_2d_fracture_dram_sharded`

**Comment by abuckTT** [comment]
I got the test to run overnight on a syseng 6U system with 32 chips. I ran 100 times and all tests passed. The problem looks 4U specific to me. 6U test also ran at 14Gbps compared to 12Gbps DRAM speed used on 4U. So at higher speeds 6U is still not failing.

I'd suggest for the to-be-deprecated 4U systems, if we see a particular module (chip) that exhibits this behaviour more frequently, we could replace that specific module. Alternatively, we could also reduce the memory speed for the 4U systems (previous tests have shown issue does not repro at 10Gbps).

**Comment by prajaramanTT** [comment]
@mikevin920 Is this still an open issue ? If not, can you please mark this closed ? Thanks.

**Comment by uaydonat** [comment]
The issue was root caused.
